### PR TITLE
Solutions to Lecture 1

### PR DESCRIPTION
**Pull request by bendo**
### Solutions for Lecture 1

cc @chshersh

### CODE REVIEW

File: src/Lecture1.hs
Comment (chshersh) [inline src/Lecture1.hs:58]: `Num` is explained in the following lectures, so you are one step ahead 😸
In here `Int -> Int -> Int` or `Integer -> Integer -> Integer` would also work, as a more specific type 🙂

File: src/Lecture1.hs
Comment (chshersh) [inline src/Lecture1.hs]: This is the correct solution 👍🏻 
I agree that it looks a bit ugly though 😕 

Another approach is to use `*` in this particular case to avoid ambiguity and explicit type annotations:

```suggestion
sumOfSquares x y = x * x + y * y
```

File: src/Lecture1.hs
Comment (chshersh) [inline src/Lecture1.hs:73]: Absolutely correct 👍🏻 
Alternatively, you can use the infix form of a function by putting it inside backticks :slightly_smiling_face: 
```suggestion
lastDigit n = abs n `mod` 10
```

File: src/Lecture1.hs
Comment (chshersh) [inline src/Lecture1.hs:106]: This is a neat implementation! 😎 

I'm already happy with it 🤗 However, it's possible to implement it more efficiently by dropping first and taking the required number of elements after that. I think it will be more efficient. But this is because I know how `drop` and `take` are implemented internally 😅 These both functions traverse the list until they reach the point of stopping.

So, for example, if you're given a string of 100 elements and you want to have a substring from index 70 to 99, your existing implementation will do 170 operations (99 + 1 for taking the entire list and then 70 for dropping the prefix) while dropping 70 first and taking the remaining 30 will require only 100 operations total.

But of course, I haven't explained the internals of `take` and `drop` in the first lecture. And you're not expected to know such internal implementation details at this point. Therefore, your implementation is pretty elegant 🆒

File: src/Lecture1.hs
Comment (chshersh) [inline src/Lecture1.hs]: As you specify the top-level type signature for this function, the compiler would be able to figure out what type `read` function has in this case due to type inference, so you can skip specifying it in the body:

```suggestion
strSum str = sum (map read (words str))
```

File: src/Lecture1.hs
Comment (chshersh) [inline src/Lecture1.hs:136]: You can write the type signature of `go` explicitly to fix the GHC warning about ambiguous types

**Review by chshersh — APPROVED** [review]
Congrats! The first homework is complete 👏🏼 
Hope you enjoyed the course so far. Keep up the great work, and feel free to ask any questions if you have 👌🏼

Also, remember to add the `hacktoberfest` topic to your repository before merging this PR if you want to participate in Hacktoberfest!

File: src/Lecture1.hs
Comment (bendo) [reply src/Lecture1.hs]: Nice solution, thanks.

File: src/Lecture1.hs
Comment (bendo) [reply src/Lecture1.hs:106]: Wow, good to know. Thanks for explanation.

**Comment by bendo** [comment]
@chshersh thank you for your feedback. I really appreciate it. I updated my solutions based on your feedback. Should I merge this pull request, or is it something what you would like to do? thanks

**Comment by chshersh** [comment]
@bendo I see that you already merged the PR. Feel free to merge whenever you like 🙂 
You can wait for my feedback to apply suggestions before merging or merge straightaway.

**Pull request merged**
